audio_core: Simplify sink interface

MerryMage opened the pull request:
We simplify the sink interface so it's just a callback now. We also do now do all audio processing in one place instead of duplicating code between the sinks.

Changes made:

* Added a proper lock-free audio FIFO.
* Simplify audio-stretching related code.
* Audio-stretching now occurs on the audio thread and not the emulation thread.
* Reduced popping upon underflow.
* Perform volume adjustment on audio thread.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/4194)
<!-- Reviewable:end -->

Dragios commented:
Sound does not initialise until I change audio device manually. And I need to reconfigure this everytime I open another game. I have yet to test on another system so can't confirm if this is a widespread issue.
Have a look on this matter.

Chemlo commented:
Opening the configuration dialog and selecting OK is enough to get audio, without changing settings. This error is logged upon selecting OK, and is not logged in a previous build without this issue.
`[   8.735348] Audio.Sink <Critical> audio_core/cubeb_sink.cpp:~CubebSink:82: Error stopping cubeb stream`

Dragios commented:
More info on this update:
Apparently this could be a Windows only issue as @tgsm tested on Linux build it is working fine all along.

Further investigation suggest that this could be MinGW only issue.
*Nightly + PR* (AppVeyor MSVC build) = Works fine
*Nightly + PR* (AppVeyor MinGW build) = Crash due to application unable to start correctly (0xc0000142)
*Canary-840* (Travis MinGW build) = Issue as mentioned above

wwylele commented:
@Dragios Could you do a double check to make sure that nightly/master doesn't have the same issue?

Dragios commented:
I have double check everything. Latest nightly does not have this issue. Same goes with Canary-839.
This issue only occurs starting from Canary-840 onwards due to this PR merged.

MerryMage commented:
Thanks for the bug report; I'll look into it.

MerryMage commented:
I've added more logging to the cubeb backend, and added a workaround for cubeb's winmm backend is used (cubeb has a known issue open where data callback occurs on init and not stream start).

Please do provide logs on with `Debug:Audio.Sink` if you're experiencing issues.
